File: services/rule-engine-service/app/api/v1/admin_notification_usage.py
"""Super-admin notification usage APIs for billing proofs."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response, status
from fastapi.params import Param
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services.notification_delivery import (
    NotificationDeliveryAuditService,
    NotificationUsageFilters,
)
from services.shared.tenant_context import TenantContext

router = APIRouter()


class NotificationUsageChannelSummary(BaseModel):
    attempted_count: int = 0
    accepted_count: int = 0
    delivered_count: int = 0
    failed_count: int = 0
    skipped_count: int = 0
    billable_count: int = 0


class NotificationUsageSummaryResponse(BaseModel):
    success: bool = True
    tenant_id: str
    month: str
    totals: NotificationUsageChannelSummary
    by_channel: dict[str, NotificationUsageChannelSummary]
    first_attempt_at: Optional[datetime] = None
    last_attempt_at: Optional[datetime] = None


class NotificationUsageLogRow(BaseModel):
    id: str
    tenant_id: Optional[str]
    attempted_at: datetime
    channel: str
    status: str
    event_type: str
    recipient_masked: str
    provider_name: str
    provider_message_id: Optional[str] = None
    rule_id: Optional[str] = None
    device_id: Optional[str] = None
    billable_units: int
    failure_code: Optional[str] = None
    failure_message: Optional[str] = None
    accepted_at: Optional[datetime] = None
    delivered_at: Optional[datetime] = None
    failed_at: Optional[datetime] = None
    metadata_json: Optional[dict[str, Any]] = None


class NotificationUsageLogsResponse(BaseModel):
    success: bool = True
    tenant_id: str
    month: str
    page: int
    page_size: int
    total: int
    data: list[NotificationUsageLogRow]


def _super_admin_context(request: Request) -> TenantContext:
    ctx = TenantContext.from_request(request)
    if not ctx.is_super_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={
                "code": "SUPER_ADMIN_REQUIRED",
                "message": "Only super admin can access notification usage APIs.",
            },
        )
    return ctx


def _build_filters(
    *,
    channel: Optional[str],
    status_value: Optional[str],
    rule_id: Optional[str],
    device_id: Optional[str],
    date_from: Optional[datetime],
    date_to: Optional[datetime],
    search: Optional[str],
) -> NotificationUsageFilters:
    def _resolve(value: Any) -> Any:
        if isinstance(value, Param):
            return value.default
        return value

    return NotificationUsageFilters(
        channel=_resolve(channel),
        status=_resolve(status_value),
        rule_id=_resolve(rule_id),
        device_id=_resolve(device_id),
        date_from=_resolve(date_from),
        date_to=_resolve(date_to),
        search=_resolve(search),
    )


@router.get("/{tenant_id}/summary", response_model=NotificationUsageSummaryResponse)
async def get_notification_usage_summary(
    tenant_id: str,
    request: Request,
    month: str = Query(..., description="Month in YYYY-MM format"),
    channel: Optional[str] = Query(None),
    status_value: Optional[str] = Query(None, alias="status"),
    rule_id: Optional[str] = Query(None),
    device_id: Optional[str] = Query(None),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    search: Optional[str] = Query(None, description="Match masked recipient or provider message id"),
    db: AsyncSession = Depends(get_db),
) -> NotificationUsageSummaryResponse:
    ctx = _super_admin_context(request)
    service = NotificationDeliveryAuditService(db, ctx)
    month_value = month.default if isinstance(month, Param) else month
    summary = await service.summarize_month(
        tenant_id=tenant_id,
        month=month_value,
        filters=_build_filters(
            channel=channel,
            status_value=status_value,
            rule_id=rule_id,
            device_id=device_id,
            date_from=date_from,
            date_to=date_to,
            search=search,
        ),
    )
    by_channel = {
        "email": NotificationUsageChannelSummary(),
        "sms": NotificationUsageChannelSummary(),
        "whatsapp": NotificationUsageChannelSummary(),
    }
    for row in summary.by_channel:
        by_channel[row.channel] = NotificationUsageChannelSummary(
            attempted_count=row.attempted_count,
            accepted_count=row.accepted_count,
            delivered_count=row.delivered_count,
            failed_count=row.failed_count,
            skipped_count=row.skipped_count,
            billable_count=row.billable_count,
        )
    return NotificationUsageSummaryResponse(
        tenant_id=tenant_id,
        month=month,
        totals=NotificationUsageChannelSummary(
            attempted_count=summary.totals.attempted_count,
            accepted_count=summary.totals.accepted_count,
            delivered_count=summary.totals.delivered_count,
            failed_count=summary.totals.failed_count,
            skipped_count=summary.totals.skipped_count,
            billable_count=summary.totals.billable_count,
        ),
        by_channel=by_channel,
        first_attempt_at=summary.totals.first_attempt_at,
        last_attempt_at=summary.totals.last_attempt_at,
    )


@router.get("/{tenant_id}/logs", response_model=NotificationUsageLogsResponse)
async def list_notification_usage_logs(
    tenant_id: str,
    request: Request,
    month: str = Query(..., description="Month in YYYY-MM format"),
    channel: Optional[str] = Query(None),
    status_value: Optional[str] = Query(None, alias="status"),
    rule_id: Optional[str] = Query(None),
    device_id: Optional[str] = Query(None),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    search: Optional[str] = Query(None, description="Match masked recipient or provider message id"),
    include_metadata: bool = Query(False),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
) -> NotificationUsageLogsResponse:
    ctx = _super_admin_context(request)
    service = NotificationDeliveryAuditService(db, ctx)
    month_value = month.default if isinstance(month, Param) else month
    page_value = int(page.default if isinstance(page, Param) else page)
    page_size_value = int(page_size.default if isinstance(page_size, Param) else page_size)
    include_metadata_value = bool(include_metadata.default if isinstance(include_metadata, Param) else include_metadata)
    log_page = await service.list_month_logs(
        tenant_id=tenant_id,
        month=month_value,
        filters=_build_filters(
            channel=channel,
            status_value=status_value,
            rule_id=rule_id,
            device_id=device_id,
            date_from=date_from,
            date_to=date_to,
            search=search,
        ),
        page=page_value,
        page_size=page_size_value,
    )
    rows = [
        NotificationUsageLogRow(
            id=row.id,
            tenant_id=row.tenant_id,
            attempted_at=row.attempted_at,
            channel=row.channel,
            status=row.status,
            event_type=row.event_type,
            recipient_masked=row.recipient_masked,
            provider_name=row.provider_name,
            provider_message_id=row.provider_message_id,
            rule_id=row.rule_id,
            device_id=row.device_id,
            billable_units=row.billable_units,
            failure_code=row.failure_code,
            failure_message=row.failure_message,
            accepted_at=row.accepted_at,
            delivered_at=row.delivered_at,
            failed_at=row.failed_at,
            metadata_json=row.metadata_json if include_metadata_value else None,
        )
        for row in log_page.rows
    ]
    return NotificationUsageLogsResponse(
        tenant_id=tenant_id,
        month=month_value,
        page=page_value,
        page_size=page_size_value,
        total=log_page.total,
        data=rows,
    )


@router.get("/{tenant_id}/export.csv")
async def export_notification_usage_csv(
    tenant_id: str,
    request: Request,
    month: str = Query(..., description="Month in YYYY-MM format"),
    channel: Optional[str] = Query(None),
    status_value: Optional[str] = Query(None, alias="status"),
    rule_id: Optional[str] = Query(None),
    device_id: Optional[str] = Query(None),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    search: Optional[str] = Query(None, description="Match masked recipient or provider message id"),
    db: AsyncSession = Depends(get_db),
) -> Response:
    ctx = _super_admin_context(request)
    service = NotificationDeliveryAuditService(db, ctx)
    month_value = month.default if isinstance(month, Param) else month
    channel_value = channel.default if isinstance(channel, Param) else channel
    status_value_resolved = status_value.default if isinstance(status_value, Param) else status_value
    generated_at = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    suffix: list[str] = [tenant_id, month_value]
    if channel_value:
        suffix.append(channel_value)
    if status_value_resolved:
        suffix.append(status_value_resolved)
    suffix.append(generated_at)
    filename = f"notification_usage_{'_'.join(suffix)}.csv"
    stream = service.stream_month_logs_csv(
        tenant_id=tenant_id,
        month=month_value,
        filters=_build_filters(
            channel=channel_value,
            status_value=status_value,
            rule_id=rule_id,
            device_id=device_id,
            date_from=date_from,
            date_to=date_to,
            search=search,
        ),
    )
    return StreamingResponse(
        content=stream,
        media_type="text/csv; charset=utf-8",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
            "X-Notification-Usage-Month": month_value,
            "X-Export-Generated-At": generated_at,
        },
    )

File: ui-web/lib/adminNotificationUsage.ts
import { apiFetch } from "./apiFetch.ts";
import { readResponseError } from "./responseError.ts";

const RULE_ENGINE_SERVICE_BASE = "/backend/rule-engine";

export type NotificationUsageChannel = "email" | "sms" | "whatsapp";
export type NotificationUsageStatus =
  | "queued"
  | "attempted"
  | "provider_accepted"
  | "delivered"
  | "failed"
  | "skipped";

export interface NotificationUsageCounts {
  attempted_count: number;
  accepted_count: number;
  delivered_count: number;
  failed_count: number;
  skipped_count: number;
  billable_count: number;
}

export interface NotificationUsageSummaryResponse {
  success: boolean;
  tenant_id: string;
  month: string;
  totals: NotificationUsageCounts;
  by_channel: Record<NotificationUsageChannel, NotificationUsageCounts>;
  first_attempt_at: string | null;
  last_attempt_at: string | null;
}

export interface NotificationUsageLogRow {
  id: string;
  tenant_id: string | null;
  attempted_at: string;
  channel: NotificationUsageChannel;
  status: NotificationUsageStatus;
  event_type: string;
  recipient_masked: string;
  provider_name: string;
  provider_message_id: string | null;
  rule_id: string | null;
  device_id: string | null;
  billable_units: number;
  failure_code: string | null;
  failure_message: string | null;
  accepted_at: string | null;
  delivered_at: string | null;
  failed_at: string | null;
  metadata_json: Record<string, unknown> | null;
}

export interface NotificationUsageLogsResponse {
  success: boolean;
  tenant_id: string;
  month: string;
  page: number;
  page_size: number;
  total: number;
  data: NotificationUsageLogRow[];
}

export interface NotificationUsageFilters {
  channel?: NotificationUsageChannel | "";
  status?: NotificationUsageStatus | "";
  ruleId?: string;
  deviceId?: string;
  search?: string;
  page?: number;
  pageSize?: number;
}

export interface NotificationSummaryCard {
  label: string;
  value: string;
  tone: "success" | "warning" | "danger" | "info" | "neutral";
}

function assertOk(response: Response): Promise<void> {
  if (response.ok) {
    return Promise.resolve();
  }
  return readResponseError(response).then((message) => Promise.reject(new Error(message)));
}

function asMonthLocal(date: Date): string {
  const year = date.getFullYear();
  const month = `${date.getMonth() + 1}`.padStart(2, "0");
  return `${year}-${month}`;
}

export function getCurrentMonthLocal(now = new Date()): string {
  return asMonthLocal(now);
}

export function buildNotificationUsageQuery(
  month: string,
  filters: NotificationUsageFilters = {},
): URLSearchParams {
  const params = new URLSearchParams();
  params.set("month", month);
  if (filters.channel) {
    params.set("channel", filters.channel);
  }
  if (filters.status) {
    params.set("status", filters.status);
  }
  if (filters.ruleId?.trim()) {
    params.set("rule_id", filters.ruleId.trim());
  }
  if (filters.deviceId?.trim()) {
    params.set("device_id", filters.deviceId.trim());
  }
  if (filters.search?.trim()) {
    params.set("search", filters.search.trim());
  }
  if (typeof filters.page === "number" && Number.isFinite(filters.page)) {
    params.set("page", `${Math.max(1, Math.floor(filters.page))}`);
  }
  if (typeof filters.pageSize === "number" && Number.isFinite(filters.pageSize)) {
    params.set("page_size", `${Math.min(500, Math.max(1, Math.floor(filters.pageSize)))}`);
  }
  return params;
}

export function buildNotificationUsageExportPath(
  tenantId: string,
  month: string,
  filters: NotificationUsageFilters = {},
): string {
  const params = buildNotificationUsageQuery(month, filters);
  return `${RULE_ENGINE_SERVICE_BASE}/api/v1/admin/notification-usage/${encodeURIComponent(tenantId)}/export.csv?${params.toString()}`;
}

export async function getNotificationUsageSummary(
  tenantId: string,
  month: string,
  filters: Omit<NotificationUsageFilters, "page" | "pageSize"> = {},
): Promise<NotificationUsageSummaryResponse> {
  const params = buildNotificationUsageQuery(month, filters);
  const response = await apiFetch(
    `${RULE_ENGINE_SERVICE_BASE}/api/v1/admin/notification-usage/${encodeURIComponent(tenantId)}/summary?${params.toString()}`,
    { cache: "no-store" },
  );
  await assertOk(response);
  return response.json() as Promise<NotificationUsageSummaryResponse>;
}

export async function getNotificationUsageLogs(
  tenantId: string,
  month: string,
  filters: NotificationUsageFilters = {},
): Promise<NotificationUsageLogsResponse> {
  const params = buildNotificationUsageQuery(month, filters);
  const response = await apiFetch(
    `${RULE_ENGINE_SERVICE_BASE}/api/v1/admin/notification-usage/${encodeURIComponent(tenantId)}/logs?${params.toString()}`,
    { cache: "no-store" },
  );
  await assertOk(response);
  return response.json() as Promise<NotificationUsageLogsResponse>;
}

export async function downloadNotificationUsageCsv(
  tenantId: string,
  month: string,
  filters: NotificationUsageFilters = {},
): Promise<void> {
  const url = buildNotificationUsageExportPath(tenantId, month, filters);
  const response = await apiFetch(url, { cache: "no-store" });
  await assertOk(response);
  const blob = await response.blob();
  const objectUrl = URL.createObjectURL(blob);
  const suffix = [tenantId, month, filters.channel || undefined, filters.status || undefined].filter(Boolean).join("_");
  const filename = `notification_usage_${suffix}.csv`;
  const anchor = document.createElement("a");
  anchor.href = objectUrl;
  anchor.download = filename;
  document.body.appendChild(anchor);
  anchor.click();
  document.body.removeChild(anchor);
  URL.revokeObjectURL(objectUrl);
}

export function buildSummaryCards(summary: NotificationUsageSummaryResponse): NotificationSummaryCard[] {
  const email = summary.by_channel.email?.billable_count ?? 0;
  const sms = summary.by_channel.sms?.billable_count ?? 0;
  const whatsapp = summary.by_channel.whatsapp?.billable_count ?? 0;
  return [
    { label: "SMS Billable", value: `${sms}`, tone: "info" },
    { label: "WhatsApp Billable", value: `${whatsapp}`, tone: "info" },
    { label: "Email Billable", value: `${email}`, tone: "info" },
    { label: "Total Billable", value: `${summary.totals.billable_count}`, tone: "success" },
    { label: "Failed", value: `${summary.totals.failed_count}`, tone: summary.totals.failed_count > 0 ? "danger" : "neutral" },
    { label: "Attempted", value: `${summary.totals.attempted_count}`, tone: "neutral" },
    { label: "Delivered", value: `${summary.totals.delivered_count}`, tone: "success" },
    { label: "Accepted", value: `${summary.totals.accepted_count}`, tone: "warning" },
  ];
}

export function buildNotificationUsageRequestKey(
  tenantId: string,
  month: string,
  filters: NotificationUsageFilters,
): string {
  return `${tenantId}:${buildNotificationUsageQuery(month, filters).toString()}`;
}

export function getNotificationUsageErrorMessage(error: unknown): string {
  if (error instanceof Error && error.message.trim()) {
    return error.message;
  }
  return "Failed to load notification usage data.";
}

export function shouldShowNotificationUsageEmptyState(payload: NotificationUsageLogsResponse | null): boolean {
  return Boolean(payload && (payload.data?.length ?? 0) === 0);
}

export function formatNotificationFailureReason(code: string | null, message: string | null): string {
  const segments = [code, message].filter(Boolean);
  if (!segments.length) {
    return "—";
  }
  const combined = segments.join(": ");
  return combined.length > 100 ? `${combined.slice(0, 100)}…` : combined;
}
